### DISCUSSION TITLE
Expanding the scope: adding a catch-all for extension methods

### DISCUSSION THREAD

**Discussion by benkoshy** [wtertinek/Linq2Acad#48] [Ideas]
The library's goal, as i understand it, is to have simplification APIs to minimise the boiler plate associated with dealing with the .net API, especially that pretaining around transactions.

**Question: is there any scope of adding general extension method PRs? to this code base**

**Context**

- Often times, I find myself writing custom extension methods for a particular project. 
- And then forgetting where it was. 
- I have sure that many others have similar issues.

If there are no objections, I can add a few here and there as needs arise. If you don't like it etc. then of course feel free to close it.

**Comment by kdub-nz** [comment]
Gist ?

**Reply by benkoshy** [reply]
Hey Kerry

I'm talking generally about extension methods - without a specific gist as an example. We see handy snippets on the swamp forum all the time. i had a project full of them.

**Comment by wtertinek** [comment]
Hi Ben,

extension methods definitely have their place. However, one has to be really careful to keep them general. Otherwise you easily build up an extensive library of special-case-extension-methods and end up not seeing the wood for the trees.

Here are the extension methods we have in Linq2Acad:

```csharp
public static bool Contains(this AttributeCollection attributes, string tag)
public static string GetValue(this AttributeCollection attributes, string tag)
public static void SetValue(this AttributeCollection attributes, string tag, string value)
public static void ClearValues(this AttributeCollection attributes)

public static void SaveData<T>(this DBObject source, string key, T data)
public static T GetData<T>(this DBObject source, string key)
public static T GetData<T>(this DBObject source, string key, bool useXData)
public static bool HasData(this DBObject source, string key)
public static bool HasData(this DBObject source, string key, bool useXData)
public static void RemoveData(this DBObject source, string key)

public static void ForEach<T>(this IEnumerable<T> elements, Action<T> action) where T : DBObject
public static IEnumerable<T> UpgradeOpen<T>(this IEnumerable<T> source) where T : DBObject
public static IEnumerable<T> DowngradeOpen<T>(this IEnumerable<T> source) where T : DBObject

public static bool Contains(this DynamicBlockReferencePropertyCollection properties, string name)
public static object GetValue(this DynamicBlockReferencePropertyCollection properties, string name)
public static void SetValue(this DynamicBlockReferencePropertyCollection properties, string name, object value)

public static void WriteLine(this Editor editor, string formatString, params object[] args)
public static PromptResult GetString(this Editor editor, string message, Func<string, bool> validate)
public static PromptResult GetString(this Editor editor, string message, Func<string, bool> validate, string errorMessage)

public static void Append(this Group group, IEnumerable<ObjectId> ids)
public static void Append(this Group group, IEnumerable<Entity> entities)
public static void Prepend(this Group group, IEnumerable<ObjectId> ids)
public static void Prepend(this Group group, IEnumerable<Entity> entities)
public static void InsertAt(this Group group, int index, IEnumerable<ObjectId> ids)
public static void InsertAt(this Group group, int index, IEnumerable<Entity> entities)
public static void Remove(this Group group, IEnumerable<ObjectId> ids)
public static void Remove(this Group group, IEnumerable<Entity> entities)
public static void RemoveAt(this Group group, int index, IEnumerable<ObjectId> ids)
public static void RemoveAt(this Group group, int index, IEnumerable<Entity> entities)

public static void Add(this LayerTableRecord layer, Entity entity)
public static void AddRange(this LayerTableRecord layer, IEnumerable<Entity> entities)
```

What you have to bear in mind with extension methods on AutoCAD types is that the implementation might require a transaction. However, beeing in the scope of a transaction cannot be guaranteed and only checked at runtime. The rest of Linq2Acad is different, because with AcadDatabase as the entry point you always have a transaction available. So (some) extension methods might contradict the design of Linq2Acad that you do not need to have any "basic knowledge" about the use of the API in mind. But I think that's a compromise you have to make.

To answer your question: yes, extension methods are wellcome, but they have to be really general. Like the ones we already have, which I believe are general.

Some rough guidelines:

* No extension methods that implement specific workflows or coding styles
* Simply combining consecutive method calls into an extension method might not be a good idea
* Type conversions might be useful (like adding a method overload that uses `IEnumerable<ObjectId>` instead of `ObjectIdCollection`)
* Adding missing standard functions on types can be useful (like `Contains`)

**Reply by wtertinek** [reply]
Of course, it is possible to start and commit a transaction in the extension method if no transaction is available. But this, in the worst case, produce a lot of overhead and is therefore not recommended. I did this with SaveData, for example (that was one of my first extension methods), but I wouldn't do it that way today (and perhaps it should be changed).

**Reply by benkoshy** [reply]
Ok will keep it generic and use the above as a guide. (btw the attributes collection extension methods i find extremely helpful.)

I did not envisage such a detailed answer, i apologise for the impost on your time.

**Reply by wtertinek** [reply]
No worries at all! I actually enjoy providing detailed responses. Sometimes they are too detailed... ;)
